File: packages/dct-sdk/src/trustScores.mjs
/**
 * DCT Trust Score System — JS port of pythonNodes/trustScores.py
 * Three signals → composite score, tier, and delegation gates.
 */
import { keccak256, toUtf8Bytes, hexlify, getBytes, isHexString } from "ethers";

/** @enum {number} */
export const AgentTier = {
  COLD: 0,
  BRONZE: 1,
  SILVER: 2,
  GOLD: 3,
};

export const TIER_NAMES = ["COLD", "BRONZE", "SILVER", "GOLD"];

const DECAY_HALF_LIFE_DAYS = 7.0;
const _W1 = 0.5;
const _W2 = 0.2;
const _W3 = 0.3;
const _S1_RAW = 0.62;
const _S1_EMA = 0.38;
const _PRIOR = 0.5;
const _S2_ABSENT_EXPECTATION_PENALTY = 0.22;
const _GOLD_MIN_DISTINCT_TOOLS = 2;
const _GOLD_DIVERSITY_MIN_EXECUTIONS = 35;

const _TIER_GATES = {
  [AgentTier.COLD]: { max_children: 1, max_depth: 1, max_spend_fraction: 0.1 },
  [AgentTier.BRONZE]: { max_children: 2, max_depth: 2, max_spend_fraction: 0.25 },
  [AgentTier.SILVER]: { max_children: 5, max_depth: 4, max_spend_fraction: 0.6 },
  [AgentTier.GOLD]: { max_children: 10, max_depth: 6, max_spend_fraction: 0.9 },
};

/**
 * @typedef {{
 *   agent_id: number,
 *   tool: string,
 *   scope_adhered: boolean,
 *   completed: boolean,
 *   spend_declared: number,
 *   spend_limit: number,
 *   latency_ms: number,
 *   timestamp: Date,
 *   response_body?: Record<string, unknown>,
 *   revocation_id?: string | null,
 * }} ExecutionEvent
 */

/**
 * @typedef {{ tool: string, validator: (body: Record<string, unknown>) => boolean }} TaskExpectation
 */

function toolMatchesExpectation(eventTool, expectationKey) {
  if (!expectationKey || !eventTool) return false;
  if (eventTool === expectationKey) return true;
  const et = eventTool.trim().toLowerCase();
  if (!(et.startsWith("0x") && et.length === 66)) return false;
  const expectedHash = keccak256(toUtf8Bytes(expectationKey.trim())).toLowerCase();
  return et === expectedHash;
}

export function resolveExpectationForEvent(e, expectations) {
  if (!expectations || typeof expectations !== "object") return null;
  for (const key of Object.keys(expectations)) {
    if (toolMatchesExpectation(e.tool, key)) return expectations[key];
  }
  return null;
}

function responseBodyHasValidatorPayload(body) {
  if (!body || typeof body !== "object") return false;
  for (const v of Object.values(body)) {
    if (v != null && v !== "" && !(Array.isArray(v) && v.length === 0) && !(typeof v === "object" && v && Object.keys(v).length === 0)) {
      return true;
    }
  }
  return false;
}

export function inferTaskCompleted(e, exp) {
  try {
    if (exp.validator(e.response_body || {})) return true;
  } catch {
    /* ignore */
  }
  if (e.revocation_id == null) return false;
  if (!e.scope_adhered) return false;
  if (e.spend_limit <= 0) return false;
  if (e.spend_declared > e.spend_limit) return false;
  if (responseBodyHasValidatorPayload(e.response_body || {})) return false;
  return true;
}

export function eventsMatchAnyExpectation(agentId, events, expectations) {
  if (!expectations || Object.keys(expectations).length === 0) return false;
  for (const ev of events) {
    if (ev.agent_id !== agentId) continue;
    if (resolveExpectationForEvent(ev, expectations) != null) return true;
  }
  return false;
}

export function signal1ScopeAdherence(agentId, events) {
  const agentEvents = events.filter((e) => e.agent_id === agentId);
  if (agentEvents.length === 0) return null;
  const adhered = agentEvents.filter((e) => e.scope_adhered).length;
  return adhered / agentEvents.length;
}

function scopeAdherenceEma(agentEvents, alpha = 0.12) {
  if (agentEvents.length === 0) return 0.5;
  const ordered = [...agentEvents].sort((a, b) => a.timestamp.getTime() - b.timestamp.getTime());
  let ema = ordered[0].scope_adhered ? 1.0 : 0.0;
  for (let i = 1; i < ordered.length; i++) {
    const x = ordered[i].scope_adhered ? 1.0 : 0.0;
    ema = alpha * x + (1.0 - alpha) * ema;
  }
  return ema;
}

export function signal1EffectiveForComposite(agentId, events) {
  const raw = signal1ScopeAdherence(agentId, events);
  if (raw == null) return null;
  const agentEvents = events.filter((e) => e.agent_id === agentId);
  const ema = scopeAdherenceEma(agentEvents);
  return _S1_RAW * raw + _S1_EMA * ema;
}

export function signal2TaskCompletion(agentId, events, expectations) {
  const agentEvents = events.filter(
    (e) => e.agent_id === agentId && resolveExpectationForEvent(e, expectations) != null
  );
  if (agentEvents.length === 0) return null;
  let completed = 0;
  for (const event of agentEvents) {
    const exp = resolveExpectationForEvent(event, expectations);
    if (exp && inferTaskCompleted(event, exp)) completed += 1;
  }
  return completed / agentEvents.length;
}

export function signal3OutcomeQuality(agentId, events, now = new Date(), expectations = null) {
  const agentEvents = events.filter((e) => e.agent_id === agentId);
  if (agentEvents.length === 0) return null;
  const k = Math.log(2) / DECAY_HALF_LIFE_DAYS;
  const weights = [];
  const scores = [];

  for (const e of agentEvents) {
    let ts = e.timestamp;
    const ageDays = Math.max((now.getTime() - ts.getTime()) / 86400000, 0);
    const weight = Math.exp(-k * ageDays);

    let spendEfficiency = e.spend_limit > 0 ? e.spend_declared / e.spend_limit : 1.0;
    spendEfficiency = Math.min(spendEfficiency, 1.0);

    let done;
    if (expectations && Object.keys(expectations).length > 0) {
      const exp = resolveExpectationForEvent(e, expectations);
      if (exp != null) done = inferTaskCompleted(e, exp);
      else done = e.completed;
    } else {
      done = e.completed;
    }

    const quality = 0.6 * (done ? 1 : 0) + 0.3 * (e.scope_adhered ? 1 : 0) + 0.1 * (1.0 - spendEfficiency);
    weights.push(weight);
    scores.push(quality);
  }

  const totalWeight = weights.reduce((a, b) => a + b, 0);
  if (totalWeight === 0) return null;
  return weights.reduce((acc, w, i) => acc + w * scores[i], 0) / totalWeight;
}

function distinctToolCount(agentEvents) {
  const set = new Set();
  for (const e of agentEvents) {
    if (e.tool) set.add(e.tool.toLowerCase());
  }
  return set.size;
}

function deriveTier(composite, executionCount, s2, agentId, expectations, agentEvents) {
  if (executionCount === 0) return AgentTier.COLD;

  if (expectations && Object.keys(expectations).length > 0 && executionCount > 0) {
    if (!eventsMatchAnyExpectation(agentId, agentEvents, expectations)) return AgentTier.BRONZE;
  }

  if (s2 != null && s2 < 0.5) return AgentTier.BRONZE;

  const distinct = distinctToolCount(agentEvents);

  if (executionCount >= 50 && composite >= 0.85) {
    if (executionCount >= _GOLD_DIVERSITY_MIN_EXECUTIONS && distinct < _GOLD_MIN_DISTINCT_TOOLS) return AgentTier.SILVER;
    return AgentTier.GOLD;
  }
  if (executionCount >= 10 && composite >= 0.7) return AgentTier.SILVER;
  return AgentTier.BRONZE;
}

/**
 * @param {number} agentId
 * @param {ExecutionEvent[]} events
 * @param {Record<string, TaskExpectation>} expectations
 * @param {Date} [now]
 */
export function computeTrustProfile(agentId, events, expectations, now = new Date()) {
  const s1 = signal1ScopeAdherence(agentId, events);
  const s1Eff = signal1EffectiveForComposite(agentId, events);
  const s2 = signal2TaskCompletion(agentId, events, expectations);
  const s3 = signal3OutcomeQuality(agentId, events, now, expectations);

  const agentEvents = events.filter((e) => e.agent_id === agentId);
  const executionCount = agentEvents.length;

  if (executionCount === 0) {
    const gates = _TIER_GATES[AgentTier.COLD];
    return {
      agent_id: agentId,
      composite_score: 0,
      tier: AgentTier.COLD,
      tier_name: TIER_NAMES[AgentTier.COLD],
      signal_1: null,
      signal_2: null,
      signal_3: null,
      execution_count: 0,
      ...gates,
    };
  }

  const s1Val = s1Eff != null ? s1Eff : _PRIOR;
  const s3Val = s3 != null ? s3 : _PRIOR;

  const covers = eventsMatchAnyExpectation(agentId, events, expectations);
  let s2Val;
  if (s2 != null) s2Val = s2;
  else if (!expectations || Object.keys(expectations).length === 0) s2Val = _PRIOR;
  else if (!covers) s2Val = _S2_ABSENT_EXPECTATION_PENALTY;
  else s2Val = _PRIOR;

  const composite = _W1 * s1Val + _W2 * s2Val + _W3 * s3Val;
  const tier = deriveTier(composite, executionCount, s2, agentId, expectations, agentEvents);
  const gates = _TIER_GATES[tier];

  return {
    agent_id: agentId,
    composite_score: composite,
    tier,
    tier_name: TIER_NAMES[tier],
    signal_1: s1,
    signal_2: s2,
    signal_3: s3,
    execution_count: executionCount,
    ...gates,
  };
}

/**
 * Decode TLSNotary attestation dict → ExecutionEvent (mirrors Python parse_tlsn_attestation).
 * @param {Record<string, unknown>} raw
 * @returns {ExecutionEvent | null}
 */
export function parseTlsnAttestation(raw) {
  try {
    const tsRaw = raw.timestamp;
    let ts;
    if (typeof tsRaw === "string") {
      ts = new Date(tsRaw.replace("Z", "+00:00"));
    } else if (typeof tsRaw === "number") {
      ts = new Date(tsRaw * 1000);
    } else {
      return null;
    }

    const status = Number(raw.status_code);
    const scopeAdhered = status >= 200 && status < 300;

    return {
      agent_id: Number(raw.agent_id),
      tool: String(raw.tool),
      scope_adhered: scopeAdhered,
      completed: false,
      spend_declared: Number(raw.spend_declared),
      spend_limit: Number(raw.spend_limit),
      latency_ms: Number(raw.latency_ms ?? 0),
      timestamp: ts,
      response_body: typeof raw.response_body === "object" && raw.response_body ? /** @type {Record<string, unknown>} */ (raw.response_body) : {},
      revocation_id: null,
    };
  } catch {
    return null;
  }
}

export function keccakToolHex(toolName) {
  return keccak256(toUtf8Bytes(toolName)).toLowerCase();
}

/** Normalize revocation id to 0x…32 bytes hex for hint maps */
export function revocationIdHex(rid) {
  if (rid == null) return null;
  if (typeof rid === "string" && isHexString(rid)) {
    const b = getBytes(rid);
    const slice = b.length <= 32 ? b : b.slice(-32);
    return hexlify(slice).toLowerCase();
  }
  return null;
}

File: pythonNodes/trustScores.py
"""
DCT Trust Score System
======================
Three signals combined into a TrustProfile that gates on-chain delegation.

Signal 1 - Scope adherence (cryptographic, from TLSNotary / on-chain events)
Signal 2 - Task completion (validator-defined, off-chain)
Signal 3 - Outcome quality (time-weighted heuristic, off-chain)
"""

from __future__ import annotations

import math
from dataclasses import dataclass, field
from datetime import datetime, timezone
from enum import IntEnum
from typing import Callable, Optional
from web3 import Web3


# ---------------------------------------------------------------------------
# Data types
# ---------------------------------------------------------------------------

@dataclass
class ExecutionEvent:
    """
    One execution record. Populated from on-chain events + off-chain log.

    scope_adhered   : from DCTEnforcer — ActionValidated / ActionRejected
    completed       : from Signal 2 validator
    response_body   : selectively revealed portion from TLSNotary attestation
    spend_declared  : USDC (6 decimals) declared in the action
    spend_limit     : USDC (6 decimals) limit in the agent's scope
    latency_ms      : wall-clock time for the action
    timestamp       : UTC time of execution
    tool            : tool name string
    agent_id        : ERC-8004 token ID
    """
    agent_id:       int
    tool:           str
    scope_adhered:  bool
    completed:      bool
    spend_declared: int
    spend_limit:    int
    latency_ms:     int
    timestamp:      datetime
    response_body:  dict = field(default_factory=dict)


@dataclass
class TaskExpectation:
    """
    Defined at delegation time by the parent agent or human root.
    The validator is the only authoritative source for Signal 2.
    """
    tool:      str
    validator: Callable[[dict], bool]


class AgentTier(IntEnum):
    COLD   = 0   # no history at all
    BRONZE = 1   # < 10 executions or composite < 0.5
    SILVER = 2   # >= 10 executions, composite >= 0.7
    GOLD   = 3   # >= 50 executions, composite >= 0.85


@dataclass
class TrustProfile:
    agent_id:            int
    composite_score:     float        # 0.0 – 1.0
    tier:                AgentTier
    signal_1:            Optional[float]   # scope adherence rate
    signal_2:            Optional[float]   # task completion rate
    signal_3:            Optional[float]   # outcome quality
    execution_count:     int
    # On-chain delegation gates derived from tier + composite
    max_children:        int          # width gate
    max_depth:           int          # depth gate
    max_spend_fraction:  float        # fraction of parent spend limit


# ---------------------------------------------------------------------------
# Signal 1 — Scope adherence  (cryptographic backing via TLSNotary)
# ---------------------------------------------------------------------------

def signal_1_scope_adherence(
    agent_id: int,
    events: list[ExecutionEvent],
) -> Optional[float]:
    """
    Fraction of executions where the agent stayed within declared scope.
    This is the only signal with cryptographic backing — treat it as ground truth.
    Returns None on cold start (no events for this agent).
    """
    agent_events = [e for e in events if e.agent_id == agent_id]
    if not agent_events:
        return None

    adhered = sum(1 for e in agent_events if e.scope_adhered)
    return adhered / len(agent_events)


# ---------------------------------------------------------------------------
# Signal 2 — Task completion  (validator-defined)
# ---------------------------------------------------------------------------

def signal_2_task_completion(
    agent_id:     int,
    events:       list[ExecutionEvent],
    expectations: dict[str, TaskExpectation],
) -> Optional[float]:
    """
    Fraction of completed tasks as judged by per-tool validators.
    Validators are defined at delegation time — never by the executing agent.

    Only events whose tool has a registered expectation are scored.
    Events with no expectation are silently skipped (not penalised).
    """
    agent_events = [
        e for e in events
        if e.agent_id == agent_id and e.tool in expectations
    ]
    if not agent_events:
        return None

    completed = 0
    for event in agent_events:
        exp = expectations[event.tool]
        try:
            if exp.validator(event.response_body):
                completed += 1
        except Exception:
            pass   # validator crash = not completed, not an error

    return completed / len(agent_events)


# ---------------------------------------------------------------------------
# Signal 3 — Outcome quality  (time-weighted heuristic)
# ---------------------------------------------------------------------------

DECAY_HALF_LIFE_DAYS = 7.0   # events older than 7 days count for ~50%


def signal_3_outcome_quality(
    agent_id: int,
    events:   list[ExecutionEvent],
    now:      Optional[datetime] = None,
) -> Optional[float]:
    """
    Time-weighted composite quality per execution.

    Quality per execution:
      0.60 — task completed (S2 completion flag on the event)
      0.30 — scope adhered
      0.10 — spend efficiency (spend_declared / spend_limit, lower = better)

    Exponential time decay: events older than DECAY_HALF_LIFE_DAYS
    count proportionally less. Recent behaviour dominates.
    """
    if now is None:
        now = datetime.now(timezone.utc)

    agent_events = [e for e in events if e.agent_id == agent_id]
    if not agent_events:
        return None

    k = math.log(2) / DECAY_HALF_LIFE_DAYS   # decay constant

    weights: list[float] = []
    scores:  list[float] = []

    for e in agent_events:
        # Normalise timezone
        ts = e.timestamp
        if ts.tzinfo is None:
            ts = ts.replace(tzinfo=timezone.utc)

        age_days = max((now - ts).total_seconds() / 86400, 0)
        weight   = math.exp(-k * age_days)

        spend_efficiency = (
            e.spend_declared / e.spend_limit
            if e.spend_limit > 0 else 1.0
        )
        spend_efficiency = min(spend_efficiency, 1.0)

        quality = (
            0.60 * float(e.completed) +
            0.30 * float(e.scope_adhered) +
            0.10 * (1.0 - spend_efficiency)
        )

        weights.append(weight)
        scores.append(quality)

    total_weight = sum(weights)
    if total_weight == 0:
        return None

    return sum(w * s for w, s in zip(weights, scores)) / total_weight


# ---------------------------------------------------------------------------
# Tier gates
# ---------------------------------------------------------------------------

_TIER_GATES: dict[AgentTier, dict] = {
    AgentTier.COLD: dict(
        max_children       = 1,
        max_depth          = 1,
        max_spend_fraction = 0.10,
    ),
    AgentTier.BRONZE: dict(
        max_children       = 2,
        max_depth          = 2,
        max_spend_fraction = 0.25,
    ),
    AgentTier.SILVER: dict(
        max_children       = 5,
        max_depth          = 4,
        max_spend_fraction = 0.60,
    ),
    AgentTier.GOLD: dict(
        max_children       = 10,
        max_depth          = 6,
        max_spend_fraction = 0.90,
    ),
}


def _derive_tier(
    composite:       float,
    execution_count: int,
    s2:              Optional[float],
) -> AgentTier:
    """
    Tier assignment rules:
    - COLD   : no executions
    - GOLD   : >= 50 executions, composite >= 0.85, s2 never failed catastrophically
    - SILVER : >= 10 executions, composite >= 0.70
    - BRONZE : everything else with data
    """
    if execution_count == 0:
        return AgentTier.COLD

    # If task completion is measurably bad, cap at BRONZE regardless of other signals
    if s2 is not None and s2 < 0.50:
        return AgentTier.BRONZE

    if execution_count >= 50 and composite >= 0.85:
        return AgentTier.GOLD
    if execution_count >= 10 and composite >= 0.70:
        return AgentTier.SILVER
    return AgentTier.BRONZE


# ---------------------------------------------------------------------------
# Combined trust profile
# ---------------------------------------------------------------------------

# Signal weights — S1 is cryptographically backed so it dominates
_W1 = 0.65   # scope adherence  (cryptographic)
_W3 = 0.35   # outcome quality  (heuristic)

# Neutral prior used when a signal has no data yet
_PRIOR = 0.5


def compute_trust_profile(
    agent_id:     int,
    events:       list[ExecutionEvent],
    expectations: dict[str, TaskExpectation],
    now:          Optional[datetime] = None,
) -> TrustProfile:
    """
    Combine all three signals into a TrustProfile.

    S1 and S3 drive the composite score.
    S2 acts as a gate — a low completion rate caps the tier at BRONZE.
    """
    s1 = signal_1_scope_adherence(agent_id, events)
    s2 = signal_2_task_completion(agent_id, events, expectations)
    s3 = signal_3_outcome_quality(agent_id, events, now)

    agent_events     = [e for e in events if e.agent_id == agent_id]
    execution_count  = len(agent_events)

    if execution_count == 0:
        return TrustProfile(
            agent_id           = agent_id,
            composite_score    = 0.0,
            tier               = AgentTier.COLD,
            signal_1           = None,
            signal_2           = None,
            signal_3           = None,
            execution_count    = 0,
            **_TIER_GATES[AgentTier.COLD],
        )

    # Fill missing signals with neutral prior — do not treat absence as failure
    s1_val = s1 if s1 is not None else _PRIOR
    s3_val = s3 if s3 is not None else _PRIOR

    composite = _W1 * s1_val + _W3 * s3_val
    tier      = _derive_tier(composite, execution_count, s2)
    gates     = _TIER_GATES[tier]

    return TrustProfile(
        agent_id          = agent_id,
        composite_score   = composite,
        tier              = tier,
        signal_1          = s1,
        signal_2          = s2,
        signal_3          = s3,
        execution_count   = execution_count,
        **gates,
    )


# ---------------------------------------------------------------------------
# TLSNotary attestation parser  (Phase 2 integration point)
# ---------------------------------------------------------------------------

def parse_tlsn_attestation(raw: dict) -> Optional[ExecutionEvent]:
    """
    Convert a decoded TLSNotary attestation into an ExecutionEvent.

    Expected attestation fields (selectively revealed by the prover):
      endpoint_hash  : keccak256 of tool name (hex string)
      status_code    : HTTP status
      response_body  : dict of selectively revealed fields
      spend_declared : int (6-decimal USDC)
      spend_limit    : int (6-decimal USDC)
      agent_id       : int
      tool           : str
      timestamp      : ISO-8601 UTC string
      latency_ms     : int

    Returns None if the attestation is malformed.
    """
    try:
        ts_raw = raw["timestamp"]
        if isinstance(ts_raw, str):
            ts = datetime.fromisoformat(ts_raw.replace("Z", "+00:00"))
        else:
            ts = datetime.fromtimestamp(ts_raw, tz=timezone.utc)

        status = int(raw["status_code"])
        scope_adhered = (200 <= status < 300)

        return ExecutionEvent(
            agent_id       = int(raw["agent_id"]),
            tool           = str(raw["tool"]),
            scope_adhered  = scope_adhered,
            completed      = False,     # Signal 2 fills this separately
            spend_declared = int(raw["spend_declared"]),
            spend_limit    = int(raw["spend_limit"]),
            latency_ms     = int(raw.get("latency_ms", 0)),
            timestamp      = ts,
            response_body  = raw.get("response_body", {}),
        )
    except (KeyError, ValueError, TypeError):
        return None


# ---------------------------------------------------------------------------
# On-chain event reader  (Phase 3 integration point)
# ---------------------------------------------------------------------------

def load_events_from_chain(
    registry_contract,          # web3.eth.Contract instance
    enforcer_contract,          # web3.eth.Contract instance
    from_block: int = 0,
) -> list[ExecutionEvent]:
    """
    Read ActionValidated and ActionRejected events from DCTEnforcer
    and produce ExecutionEvent records with scope_adhered set correctly.

    This is a read-only operation — no transactions.
    """
    events: list[ExecutionEvent] = []

    try:
        validated = enforcer_contract.events.ActionValidated.get_logs(
            fromBlock=from_block
        )
        for log in validated:
            a = log["args"]
            events.append(ExecutionEvent(
                agent_id       = a["agentTokenId"],
                tool           = Web3.to_hex(a["toolHash"]),
                scope_adhered  = True,
                completed      = False,
                spend_declared = a["spendAmount"],
                spend_limit    = 0,    # fetch from scopeCommitments if needed
                latency_ms     = 0,
                timestamp      = datetime.now(timezone.utc),
                response_body  = {},
            ))

        rejected = enforcer_contract.events.ActionRejected.get_logs(
            fromBlock=from_block
        )
        for log in rejected:
            a = log["args"]
            events.append(ExecutionEvent(
                agent_id       = a["agentTokenId"],
                tool           = Web3.to_hex(a["toolHash"]) if "toolHash" in a else "",
                scope_adhered  = False,
                completed      = False,
                spend_declared = 0,
                spend_limit    = 0,
                latency_ms     = 0,
                timestamp      = datetime.now(timezone.utc),
                response_body  = {},
            ))
    except Exception as exc:
        print(f"[load_events_from_chain] warning: {exc}")

    return events